applications/FluidDynamicsApplication/custom_processes/shock_capturing_entropy_viscosity_process.cpp

```diff
@@ -0,0 +1,390 @@
+    KRATOS_ERROR_IF(missing_density) << "Missing DENSITY variable from one or more nodes" << std::endl;
+    KRATOS_ERROR_IF(missing_pressure) << "Missing PRESSURE variable from one or more nodes" << std::endl;
+    KRATOS_ERROR_IF(missing_temperature) << "Missing TEMPERATURE variable from one or more nodes" << std::endl;
+    KRATOS_ERROR_IF(missing_velocity) << "Missing VELOCITY variable from one or more nodes" << std::endl;
+
+    if(mrModelPart.ElementsBegin() != mrModelPart.ElementsEnd())
+    {
+        KRATOS_ERROR_IF_NOT(mrModelPart.ElementsBegin()->GetProperties().Has(HEAT_CAPACITY_RATIO))
+            << "Variable HEAT_CAPACITY_RATIO missing from elemental properties." << std::endl;
+
+#ifdef KRATOS_DEBUG
+        const double first_gamma = mrModelPart.ElementsBegin()->GetProperties().GetValue(HEAT_CAPACITY_RATIO);
+
+        const bool non_uniform_gamma =
+        block_for_each<OrReduction>(mrModelPart.Elements(), [&first_gamma](Element& r_element)
+        {
+            static constexpr double tolerance = 1e-8;
+            const double this_gamma = r_element.GetProperties().GetValue(HEAT_CAPACITY_RATIO);
+            return std::fabs(first_gamma - this_gamma) > tolerance;
+        });
+
+        KRATOS_ERROR_IF(non_uniform_gamma)
+            << "HEAT_CAPACITY_RATIO is not constant in the domain" << std::endl;
+#endif
+    }
+
+    return err_code;
+}
+
+
+const Parameters ShockCapturingEntropyViscosityProcess::GetDefaultParameters() const
+{
+    return Parameters(R"(
+    {
+        "model_part_name" : "",
+        "calculate_nodal_area_at_each_step" : false,
+        "entropy_constant"  : 0.0,
+        "energy_constant"   : 0.0,
+        "artificial_mass_viscosity_Prandtl"   : 0.1,
+        "artificial_conductivity_Prandtl"     : 0.1
+    }
+    )");
+}
+
+
+double ShockCapturingEntropyViscosityProcess::TotalDerivativeUtil::
+Divergence(const Matrix& rShapeFunGradients, const Matrix& rNodalValues)
+{
+    KRATOS_TRY
+
+    KRATOS_DEBUG_ERROR_IF(rShapeFunGradients.size1() != rNodalValues.size1())
+        << "Matrices rShapeFunGradients and rNodalValues must be the same size (Size1: "
+        << rShapeFunGradients.size1() << " != " << rNodalValues.size1() <<")" << std::endl;
+    KRATOS_DEBUG_ERROR_IF(rShapeFunGradients.size2() != rNodalValues.size2())
+        << "Matrices rShapeFunGradients and rNodalValues must be the same size (Size2: "
+        << rShapeFunGradients.size2() << " != " << rNodalValues.size2() <<")" << std::endl;
+
+    double divergence = 0.0;
+    for(std::size_t i=0; i<rShapeFunGradients.size1(); ++i)
+    {
+        for(std::size_t j=0; j<rShapeFunGradients.size2(); ++j)
+        {
+            divergence += rShapeFunGradients(i,j) * rNodalValues(i,j);
+        }
+    }
+    return divergence;
+
+    KRATOS_CATCH("")
+}
+
+
+void ShockCapturingEntropyViscosityProcess::ExecuteInitializeSolutionStep()
+{
+    if(mFirstTimeStep)
```

Comment:
```suggestion
    if(mIsInitialized)
```
I think this is a better name.